soccer/src/soccer/coach/coach_node.cpp

```diff
@@ -0,0 +1,140 @@
+#include "coach_node.hpp"
+
+CoachNode::CoachNode(const rclcpp::NodeOptions& options) : Node("coach_node", options) {
+    coach_pub_ =
+        this->create_publisher<rj_msgs::msg::CoachStateInterpretation>("/strategy/coach", 10);
+    play_state_change_timer_ =
+        this->create_wall_timer(100ms, [this]() { check_for_play_state_change(); });
+
+    play_state_sub_ = this->create_subscription<rj_msgs::msg::PlayState>(
+        "/referee/play_state", 10,
+        [this](rj_msgs::msg::PlayState::SharedPtr msg) { play_state_callback(msg); });
+
+    world_state_sub_ = this->create_subscription<rj_msgs::msg::WorldState>(
+        "/vision_filter/world_state", 10,
+        [this](rj_msgs::msg::WorldState::SharedPtr msg) { world_state_callback(msg); });
+
+    // initialize all of the robot status subscriptions
+    // initialize our robots
+    for (int i = 0; i < kRobotsPerTeam; i++) {
+        robot_status_subs_[i] = this->create_subscription<rj_msgs::msg::RobotStatus>(
+            fmt::format("/radio/robot_status/robot_{}", i), 10,
+            [this](const rj_msgs::msg::RobotStatus::SharedPtr msg) {
+                ball_sense_callback(msg, true);
+            });
+    }
+
+    // initialize their robots
+    for (int i = kRobotsPerTeam; i < kRobotsPerTeam * 2; i++) {
+        robot_status_subs_[i] = this->create_subscription<rj_msgs::msg::RobotStatus>(
+            fmt::format("/radio/robot_status/robot_{}", i), 10,
+            [this](rj_msgs::msg::RobotStatus::SharedPtr msg) { ball_sense_callback(msg, false); });
+    }
+
+    current_play_state_.state = PlayState::State::Halt;
+    current_play_state_.restart = PlayState::Restart::Kickoff;
+    current_play_state_.our_restart = true;
+
+    rj_geometry_msgs::msg::Point temp_point;
+    temp_point.x = -1;
+    temp_point.y = -1;
+    current_play_state_.placement_point = temp_point;
+}
+
+void CoachNode::play_state_callback(rj_msgs::msg::PlayState::SharedPtr msg) {
+    current_play_state_ = *msg;
+    play_state_has_changed_ = true;
+}
+
+void CoachNode::world_state_callback(rj_msgs::msg::WorldState::SharedPtr msg) {
+    // EDGE-CASE NOTE: If robots from both teams are bordering the ball possession will likely
+    // switch repeatedly
+    if (!possessing_) {
+        for (rj_msgs::msg::RobotState robot_state : msg->our_robots) {
+            // There definitely has to be a better way, but this works...
```

Comment:
there is, see the static `nearly_equals` here: https://github.com/RoboJackets/robocup-software/blob/efc4f3ae3540917e241a669a40cd997433446ed4/rj_geometry/include/rj_geometry/point.hpp#L400

or the non-static version here: https://github.com/RoboJackets/robocup-software/blob/efc4f3ae3540917e241a669a40cd997433446ed4/rj_geometry/testing/point_test.cpp#L205

Reply:
(if I'm not mistaken you should be able to just use robot_state.pose.position as an rj_geometry Point. maybe there is some conversion function I'm forgetting)

Reply:
I don't think we can do nearly_equals here because the robot position is its centroid which should never be nearly equal to the ball's location.

I'm also not entirely sure why, but even though both the robot's position and the ball's position are rj_geometry::points the rj_geometry point seems to be different than the one in the message so I had to make that ugly initializer.

Reply:
oh you can just change the tolerance on nearly equals to be the robot radius (or whatever threshold you had already)